File: linux/app/gbw/services/source_search.py
from __future__ import annotations

import html
import logging
import re
import statistics
import unicodedata
import urllib.parse
import urllib.request
from difflib import SequenceMatcher

from gbw.core.runner import CommandRunner, CommandError
from gbw.core.system import tool_path
from gbw.models import SourceCandidate


def norm_text(value: str) -> str:
    s = unicodedata.normalize("NFKD", value or "").encode("ascii", "ignore").decode("ascii")
    return re.sub(r"[^a-z0-9]+", " ", s.lower()).strip()


def text_similarity(a: str, b: str) -> float:
    aa, bb = norm_text(a), norm_text(b)
    return SequenceMatcher(None, aa, bb).ratio() if aa and bb else 0.0


TITLE_STOPWORDS = {
    "the", "a", "an", "of", "and", "or", "to", "in", "on", "at", "for", "from", "with",
    "feat", "ft", "featuring", "de", "da", "do", "das", "dos", "e", "o", "os", "as", "um", "uma",
}


def _meaningful_tokens(value: str) -> list[str]:
    tokens = norm_text(value).split()
    meaningful = [tok for tok in tokens if tok not in TITLE_STOPWORDS]
    return meaningful or tokens


def title_matches_song(song: str, title: str) -> bool:
    """Gate eliminatório: primeiro prova que o resultado é a música pedida.

    - títulos de uma palavra exigem o token exato (Enemy != Eclipse/Enemies);
    - títulos com várias palavras exigem todos os tokens significativos, em qualquer ordem;
    - descritores extras como Official Audio/Live/Remix não impedem a correspondência.
    """
    wanted = _meaningful_tokens(song)
    got = set(norm_text(title).split())
    if not wanted or not got:
        return False
    if len(wanted) == 1:
        return wanted[0] in got
    return all(tok in got for tok in wanted)


def artist_matches_request(artist: str, title: str, uploader: str) -> bool:
    """Confirma o artista quando ele foi informado, sem exigir um uploader idêntico.

    Aceita o nome do artista no próprio título (comum em uploads de gravadoras/Topic)
    ou uma boa cobertura de tokens no uploader/título combinado.
    """
    wanted = _meaningful_tokens(artist)
    if not wanted:
        return True
    combined_tokens = set(norm_text(f"{title} {uploader}").split())
    if all(tok in combined_tokens for tok in wanted):
        return True
    uploader_n = norm_text(uploader)
    artist_n = norm_text(artist)
    if uploader_n and (artist_n in uploader_n or uploader_n in artist_n):
        return True
    coverage = sum(1 for tok in wanted if tok in combined_tokens) / max(1, len(wanted))
    return coverage >= 0.67


def duration_label(seconds: float | int | None) -> str:
    try:
        sec = int(round(float(seconds or 0)))
    except Exception:
        return "duração desconhecida"
    if sec <= 0:
        return "duração desconhecida"
    return f"{sec // 60}:{sec % 60:02d}"


class SourceSearchService:
    """Busca e ranqueia fontes preservando qualidade e evitando previews/truncamentos.

    Regras importantes:
    - lossless/download original têm prioridade estrutural;
    - formatos explicitamente marcados como preview nunca são recomendados;
    - durações muito divergentes da mediana dos candidatos equivalentes sofrem penalidade;
    - a seleção registra um format_id real quando possível, para baixar exatamente o formato inspecionado.
    """

    def __init__(self, runner: CommandRunner, logger: logging.Logger):
        self.runner = runner
        self.logger = logger

    def ytdlp_common(self) -> list[str]:
        args = ["--ignore-config"]
        deno = tool_path("deno")
        if deno:
            args += ["--js-runtimes", f"deno:{deno}"]
        return args

    def _flat_search(self, prefix: str, query: str, count: int) -> list[dict]:
        ytdlp = tool_path("yt-dlp")
        if not ytdlp:
            raise CommandError("A pesquisa online não está disponível. Verifique a página Sistema.")
        data = self.runner.capture_json([
            ytdlp, *self.ytdlp_common(), "--no-warnings", "--skip-download", "--flat-playlist",
            "--dump-single-json", f"{prefix}{count}:{query}"
        ], timeout=120)
        return [e for e in (data.get("entries") or []) if isinstance(e, dict)]

    def _entry_url(self, entry: dict, source: str) -> str:
        url = entry.get("webpage_url") or entry.get("original_url") or entry.get("url") or ""
        if source == "YouTube" and url and not str(url).startswith("http"):
            vid = entry.get("id") or url
            return f"https://www.youtube.com/watch?v={vid}"
        return str(url)

    def inspect_url(self, url: str) -> dict:
        ytdlp = tool_path("yt-dlp")
        if not ytdlp:
            raise CommandError("A pesquisa online não está disponível. Verifique a página Sistema.")
        return self.runner.capture_json([
            ytdlp, *self.ytdlp_common(), "--no-warnings", "--skip-download", "--no-playlist", "--dump-single-json", url
        ], timeout=150)

    def _bandcamp_urls(self, query: str, limit: int) -> list[str]:
        req = urllib.request.Request(
            "https://bandcamp.com/search?q=" + urllib.parse.quote_plus(query),
            headers={"User-Agent": "Mozilla/5.0 GuitarBackingWizard/5.23"},
        )
        try:
            with urllib.request.urlopen(req, timeout=20) as r:
                body = r.read().decode("utf-8", errors="replace")
        except Exception as exc:
            self.logger.debug("Bandcamp indisponível: %s", exc)
            return []
        body = html.unescape(body)
        urls = re.findall(r"https?://[^\"'<> ]+\.bandcamp\.com/track/[^\"'<>?&# ]+", body, flags=re.I)
        out, seen = [], set()
        for u in urls:
            u = u.rstrip("/.,)")
            if u not in seen:
                seen.add(u)
                out.append(u)
            if len(out) >= limit:
                break
        return out

    @staticmethod
    def _is_preview_format(fmt: dict) -> bool:
        blob = " ".join(str(fmt.get(k) or "") for k in ("format_id", "format", "format_note", "url")).lower()
        return any(token in blob for token in ("preview", "sample", "excerpt", "snippet"))

    @classmethod
    def _quality(cls, info: dict) -> tuple[str, str, int, bool]:
        formats = [
            f for f in (info.get("formats") or [])
            if isinstance(f, dict) and f.get("vcodec") in (None, "none") and str(f.get("acodec") or "none") != "none"
        ]
        non_preview = [f for f in formats if not cls._is_preview_format(f)]
        preview_only = bool(formats) and not non_preview
        usable = non_preview or formats

        lossless_exts = {"flac", "wav", "alac", "ape", "aiff", "aif"}
        lossless, originals = [], []
        for f in usable:
            ext = str(f.get("ext") or "").lower()
            acodec = str(f.get("acodec") or "").lower()
            fid = str(f.get("format_id") or "")
            note = str(f.get("format_note") or "").lower()
            if ext in lossless_exts or acodec.startswith("pcm") or acodec in {"flac", "alac"}:
                lossless.append(f)
            if fid == "download" or "original" in note:
                originals.append(f)

        if preview_only:
            f = max(usable, key=lambda x: (x.get("quality") or -99, x.get("abr") or x.get("tbr") or 0)) if usable else {}
            ext = str(f.get("ext") or "áudio").upper()
            return f"PREVIEW/TRUNCADO {ext}", str(f.get("format_id") or "bestaudio/best"), -60, True

        if lossless:
            pref = {"flac": 0, "alac": 1, "wav": 2, "aiff": 3, "aif": 3, "ape": 4}
            lossless.sort(key=lambda f: (pref.get(str(f.get("ext") or "").lower(), 9), -(f.get("filesize") or 0)))
            f = lossless[0]
            return f"LOSSLESS {str(f.get('ext') or '').upper()}", str(f.get("format_id") or "bestaudio/best"), 40, False

        if originals:
            f = originals[0]
            return f"download ORIGINAL ({str(f.get('ext') or 'arquivo').upper()})", str(f.get("format_id") or "download"), 32, False

        if usable:
            f = max(usable, key=lambda x: (x.get("quality") or -99, x.get("abr") or x.get("tbr") or 0))
            br = f.get("abr") or f.get("tbr")
            detail = f"stream {str(f.get('ext') or 'áudio').upper()}" + (f" ~{int(br)} kbps" if isinstance(br, (int, float)) and br > 0 else "")
            return detail, str(f.get("format_id") or "bestaudio/best"), 0, False

        return "áudio disponível", "bestaudio/best", 0, False

    def _candidate(self, info: dict, source: str, url: str, artist: str, song: str) -> SourceCandidate:
        title = str(info.get("track") or info.get("title") or "")
        uploader = str(info.get("artist") or info.get("uploader") or "")
        artists = info.get("artists")
        if isinstance(artists, list) and artists:
            uploader = ", ".join(str(x) for x in artists if x)

        quality, fmt, bonus, preview_only = self._quality(info)
        title_n, uploader_n, req_n, artist_n = map(norm_text, (title, uploader, song, artist))
        song_sim, artist_sim = text_similarity(song, title), text_similarity(artist, uploader)
        if req_n and req_n in title_n:
            song_sim = max(song_sim, .95)
        if artist_n and artist_n in uploader_n:
            artist_sim = max(artist_sim, .95)

        base = {"Bandcamp": 45, "SoundCloud": 48, "YouTube": 50}.get(source, 45)
        score = base + bonus + int(song_sim * 12) + int(artist_sim * 10)
        bad = ("cover", "karaoke", "reaction", "tutorial", "slowed", "sped up", "nightcore", "8d", "live")
        for token in bad:
            if token in title_n and token not in req_n:
                score -= 18 if token != "live" else 10
        if "remix" in title_n and "remix" not in req_n:
            score -= 14

        duration = float(info.get("duration") or 0.0)
        short_title_exempt = any(tok in req_n for tok in ("intro", "interlude", "outro", "short"))
        duration_warning = bool(duration and duration < 60 and not short_title_exempt)
        if duration_warning:
            score -= 30

        official = False
        reason: list[str] = []
        if bonus >= 40:
            reason.append("áudio de boa qualidade")
        elif bonus >= 30:
            reason.append("fonte direta")
        if preview_only:
            reason.append("trecho curto")
            score = min(score, 5)
        elif duration_warning:
            reason.append(f"duração curta ({duration_label(duration)})")

        if source == "YouTube":
            if "topic" in uploader_n or "official audio" in title_n or "official music" in title_n:
                score += 10
                official = True
                reason.append("canal oficial")
            elif artist_sim >= .82:
                score += 6
                reason.append("canal compatível")
        elif artist_sim >= .80:
            score += 6
            official = True
            reason.append("perfil compatível")
        if song_sim >= .82:
            reason.append("título compatível")
        if duration > 0 and not preview_only:
            reason.append(f"duração {duration_label(duration)}")

        return SourceCandidate(
            score=max(0, min(100, score)), source=source, quality=quality, title=title, uploader=uploader,
            url=url, format_id=fmt, official=official,
            reason=", ".join(reason) or "melhor resultado encontrado",
            quality_bonus=bonus, duration=duration, preview_only=preview_only, duration_warning=duration_warning,
        )

    @staticmethod
    def _candidate_matches_request(candidate: SourceCandidate, artist: str, song: str) -> bool:
        if not title_matches_song(song, candidate.title):
            return False
        if artist.strip() and not artist_matches_request(artist, candidate.title, candidate.uploader):
            return False
        return True

    def _apply_duration_consensus(self, candidates: list[SourceCandidate], song: str) -> None:
        """Penaliza versões truncadas usando a mediana das correspondências plausíveis.

        Isso resolve casos em que um serviço retorna um preview de ~30 s ao lado de uploads completos
        de 3-5 min. A regra só entra quando há pelo menos duas referências de duração plausíveis.
        """
        plausible = [
            c for c in candidates
            if not c.preview_only and c.duration >= 60 and text_similarity(song, c.title) >= .72
        ]
        if len(plausible) < 2:
            return
        med = statistics.median(c.duration for c in plausible)
        if med <= 0:
            return
        for c in candidates:
            if c.preview_only or c.duration <= 0:
                continue
            ratio = c.duration / med
            if ratio < .68:
                c.score = max(0, c.score - 35)
                c.duration_warning = True
                c.reason += f", duração muito menor que versões equivalentes (mediana {duration_label(med)})"
            elif ratio > 1.55:
                c.score = max(0, c.score - 15)
                c.duration_warning = True
                c.reason += f", duração muito maior que versões equivalentes (mediana {duration_label(med)})"
            elif .92 <= ratio <= 1.08:
                c.score = min(100, c.score + 3)

    def search(self, artist: str, song: str, depth: str = "Robusta") -> list[SourceCandidate]:
        if not song.strip():
            raise ValueError("Informe o nome da música.")
        query = " ".join(x for x in (artist.strip(), song.strip()) if x)
        max_bc, max_search, max_inspect = (6, 9, 8) if depth == "Máxima" else (4, 6, 5)
        candidates: list[SourceCandidate] = []

        for url in self._bandcamp_urls(query, max_bc):
            try:
                candidate = self._candidate(self.inspect_url(url), "Bandcamp", url, artist, song)
                if self._candidate_matches_request(candidate, artist, song):
                    candidates.append(candidate)
                else:
                    self.logger.debug("Bandcamp descartado por não corresponder à música/artista: %s — %s", candidate.uploader, candidate.title)
            except Exception as exc:
                self.logger.debug("Bandcamp candidato ignorado: %s", exc)

        searches = [
            ("SoundCloud", "scsearch", query),
            ("YouTube", "ytsearch", f"{query} official audio"),
            ("YouTube", "ytsearch", f"{query} Topic"),
        ]
        seen: set[str] = {c.url for c in candidates}
        for source, prefix, q in searches:
            try:
                entries = self._flat_search(prefix, q, max_search)
            except Exception as exc:
                self.logger.debug("Busca %s indisponível: %s", source, exc)
                continue
            inspected = 0
            for entry in entries:
                if inspected >= max_inspect:
                    break
                flat_title = str(entry.get("title") or entry.get("track") or "")
                # Evita gastar inspeções com resultados obviamente de outra música.
                if flat_title and not title_matches_song(song, flat_title):
                    self.logger.debug("%s descartado no gate de título: %s", source, flat_title)
                    continue
                url = self._entry_url(entry, source)
                if not url or url in seen:
                    continue
                seen.add(url)
                try:
                    candidate = self._candidate(self.inspect_url(url), source, url, artist, song)
                    if not self._candidate_matches_request(candidate, artist, song):
                        self.logger.debug("%s descartado por não corresponder à música/artista: %s — %s", source, candidate.uploader, candidate.title)
                        continue
                    candidates.append(candidate)
                    inspected += 1
                except Exception as exc:
                    self.logger.debug("%s candidato ignorado: %s", source, exc)

        self._apply_duration_consensus(candidates, song)

        # Previews ficam visíveis somente no fim para diagnóstico, nunca como recomendação automática.
        candidates.sort(key=lambda c: (not c.preview_only, c.score, c.quality_bonus, c.official), reverse=True)
        safe = [c for c in candidates if not c.preview_only]
        rejected = [c for c in candidates if c.preview_only]
        return (safe + rejected)[:15]

File: android/app/src/main/java/com/gbw/android/domain/SourceSearch.kt
package com.gbw.android.domain

import java.text.Normalizer
import java.util.Locale
import kotlin.math.max
import kotlin.math.min

enum class SourceProvider(val publicLabel: String) {
    BANDCAMP("Bandcamp"),
    SOUNDCLOUD("SoundCloud"),
    YOUTUBE("YouTube"),
    OTHER("Outra"),
}

enum class SourceSearchDepth {
    ROBUST,
    MAXIMUM,
}

data class SourceSearchRequest(
    val artist: String,
    val song: String,
    val depth: SourceSearchDepth = SourceSearchDepth.ROBUST,
)

data class SourceCandidateDraft(
    val provider: SourceProvider,
    val title: String,
    val uploader: String = "",
    val url: String,
    val quality: String = "áudio disponível",
    val formatId: String = "",
    val qualityBonus: Int = 0,
    val durationSeconds: Double = 0.0,
    val previewOnly: Boolean = false,
    val officialSignal: Boolean = false,
    val automaticDownloadSupported: Boolean = false,
)

data class SourceSearchLink(
    val label: String,
    val url: String,
)

object SourceSearchLinks {
    fun forRequest(request: SourceSearchRequest): List<SourceSearchLink> {
        val query = listOf(request.artist.trim(), request.song.trim())
            .filter { it.isNotBlank() }
            .joinToString(" ")
        if (query.isBlank()) return emptyList()
        val encoded = java.net.URLEncoder.encode(query, java.nio.charset.StandardCharsets.UTF_8.name())
        return listOf(
            SourceSearchLink("Pesquisar no YouTube", "https://www.youtube.com/results?search_query=$encoded"),
            SourceSearchLink("Pesquisar no SoundCloud", "https://soundcloud.com/search/sounds?q=$encoded"),
            SourceSearchLink("Pesquisar no Bandcamp", "https://bandcamp.com/search?q=$encoded"),
        )
    }
}

data class RankedSourceCandidate(
    val provider: SourceProvider,
    val title: String,
    val uploader: String,
    val url: String,
    val quality: String,
    val formatId: String,
    val qualityBonus: Int,
    val durationSeconds: Double,
    val previewOnly: Boolean,
    val durationWarning: Boolean,
    val official: Boolean,
    val score: Int,
    val reason: String,
    val automaticDownloadSupported: Boolean = true,
)

interface SourceSearchProviderClient {
    suspend fun search(request: SourceSearchRequest): List<SourceCandidateDraft>
}

/**
 * Stable Android port of the Linux 5.23 source-selection contract.
 *
 * Network/provider extraction is intentionally kept outside this object. That
 * allows provider implementations to change without changing the ranking and
 * identity rules already homologated by the desktop baseline.
 */
object SourceSearchRules {
    private val stopWords = setOf(
        "the", "a", "an", "of", "and", "or", "to", "in", "on", "at", "for", "from", "with",
        "feat", "ft", "featuring", "de", "da", "do", "das", "dos", "e", "o", "os", "as", "um", "uma",
    )

    private val undesirableDescriptors = mapOf(
        "cover" to 18,
        "karaoke" to 18,
        "reaction" to 18,
        "tutorial" to 18,
        "slowed" to 18,
        "sped" to 18,
        "nightcore" to 18,
        "8d" to 18,
        "live" to 10,
    )

    fun normalize(value: String): String {
        val ascii = Normalizer.normalize(value, Normalizer.Form.NFD)
            .replace(Regex("\\p{Mn}+"), "")
            .lowercase(Locale.ROOT)
        return ascii.replace(Regex("[^a-z0-9]+"), " ").trim()
    }

    fun meaningfulTokens(value: String): List<String> {
        val tokens = normalize(value).split(' ').filter { it.isNotBlank() }
        val meaningful = tokens.filterNot { it in stopWords }
        return meaningful.ifEmpty { tokens }
    }

    fun titleMatchesSong(song: String, title: String): Boolean {
        val wanted = meaningfulTokens(song)
        val got = normalize(title).split(' ').filter { it.isNotBlank() }.toSet()
        if (wanted.isEmpty() || got.isEmpty()) return false
        return if (wanted.size == 1) wanted.first() in got else wanted.all { it in got }
    }

    fun artistMatchesRequest(artist: String, title: String, uploader: String): Boolean {
        val wanted = meaningfulTokens(artist)
        if (wanted.isEmpty()) return true

        val combinedTokens = normalize("$title $uploader").split(' ').filter { it.isNotBlank() }.toSet()
        if (wanted.all { it in combinedTokens }) return true

        val uploaderN = normalize(uploader)
        val artistN = normalize(artist)
        if (uploaderN.isNotBlank() && (artistN in uploaderN || uploaderN in artistN)) return true

        val coverage = wanted.count { it in combinedTokens }.toDouble() / max(1, wanted.size)
        return coverage >= 0.67
    }

    fun textSimilarity(a: String, b: String): Double {
        val aa = normalize(a)
        val bb = normalize(b)
        if (aa.isBlank() || bb.isBlank()) return 0.0
        if (aa == bb) return 1.0
        if (aa in bb || bb in aa) {
            val shorter = min(aa.length, bb.length).toDouble()
            val longer = max(aa.length, bb.length).toDouble()
            return (0.85 + 0.15 * (shorter / longer)).coerceAtMost(0.99)
        }

        val distance = levenshtein(aa, bb)
        return (1.0 - distance.toDouble() / max(aa.length, bb.length).toDouble()).coerceIn(0.0, 1.0)
    }

    fun durationLabel(seconds: Double): String {
        val rounded = seconds.toInt()
        if (rounded <= 0) return "duração desconhecida"
        return "${rounded / 60}:${(rounded % 60).toString().padStart(2, '0')}"
    }

    fun rank(
        request: SourceSearchRequest,
        drafts: List<SourceCandidateDraft>,
        limit: Int = 15,
    ): List<RankedSourceCandidate> {
        val accepted = drafts
            .filter { titleMatchesSong(request.song, it.title) }
            .filter { request.artist.isBlank() || artistMatchesRequest(request.artist, it.title, it.uploader) }
            .distinctBy { it.url }
            .map { score(request, it) }
            .toMutableList()

        applyDurationConsensus(accepted, request.song)

        return accepted
            .sortedWith(
                compareByDescending<RankedSourceCandidate> { !it.previewOnly }
                    .thenByDescending { it.score }
                    .thenByDescending { it.qualityBonus }
                    .thenByDescending { it.official }
                    .thenBy { it.provider.publicLabel }
                    .thenBy { it.title },
            )
            .take(limit.coerceAtLeast(1))
    }

    private fun score(request: SourceSearchRequest, draft: SourceCandidateDraft): RankedSourceCandidate {
        val titleN = normalize(draft.title)
        val requestN = normalize(request.song)
        val uploaderN = normalize(draft.uploader)

        var songSimilarity = textSimilarity(request.song, draft.title)
        var artistSimilarity = textSimilarity(request.artist, draft.uploader)
        if (requestN.isNotBlank() && requestN in titleN) songSimilarity = max(songSimilarity, 0.95)
        val artistN = normalize(request.artist)
        if (artistN.isNotBlank() && artistN in uploaderN) artistSimilarity = max(artistSimilarity, 0.95)

        var score = when (draft.provider) {
            SourceProvider.BANDCAMP -> 45
            SourceProvider.SOUNDCLOUD -> 48
            SourceProvider.YOUTUBE -> 50
            SourceProvider.OTHER -> 45
        }
        score += draft.qualityBonus
        score += (songSimilarity * 12).toInt()
        score += (artistSimilarity * 10).toInt()

        undesirableDescriptors.forEach { (token, penalty) ->
            if (token in titleN && token !in requestN) score -= penalty
        }
        if ("remix" in titleN && "remix" !in requestN) score -= 14

        val shortTitleExempt = listOf("intro", "interlude", "outro", "short").any { it in requestN }
        var durationWarning = draft.durationSeconds in 0.000001..59.999999 && !shortTitleExempt
        if (durationWarning) score -= 30

        var official = draft.officialSignal
        val reasons = mutableListOf<String>()

        when {
            draft.qualityBonus >= 40 -> reasons += "áudio de boa qualidade"
            draft.qualityBonus >= 30 -> reasons += "fonte direta"
        }

        if (draft.previewOnly) {
            reasons += "trecho curto"
            score = min(score, 5)
        } else if (durationWarning) {
            reasons += "duração curta (${durationLabel(draft.durationSeconds)})"
        }

        when (draft.provider) {
            SourceProvider.YOUTUBE -> {
                if (draft.officialSignal || "topic" in uploaderN || "official audio" in titleN || "official music" in titleN) {
                    score += 10
                    official = true
                    reasons += "canal oficial"
                } else if (artistSimilarity >= 0.82) {
                    score += 6
                    reasons += "canal compatível"
                }
            }
            SourceProvider.BANDCAMP, SourceProvider.SOUNDCLOUD, SourceProvider.OTHER -> {
                if (draft.officialSignal || artistSimilarity >= 0.80) {
                    score += 6
                    official = true
                    reasons += "perfil compatível"
                }
            }
        }

        if (songSimilarity >= 0.82) reasons += "título compatível"
        if (draft.durationSeconds > 0 && !draft.previewOnly) {
            reasons += "duração ${durationLabel(draft.durationSeconds)}"
        }

        return RankedSourceCandidate(
            provider = draft.provider,
            title = draft.title,
            uploader = draft.uploader,
            url = draft.url,
            quality = draft.quality,
            formatId = draft.formatId,
            qualityBonus = draft.qualityBonus,
            durationSeconds = draft.durationSeconds,
            previewOnly = draft.previewOnly,
            durationWarning = durationWarning,
            official = official,
            score = (if (draft.previewOnly) min(score, 5) else score).coerceIn(0, 100),
            reason = reasons.joinToString(", ").ifBlank { "melhor resultado encontrado" },
            automaticDownloadSupported = draft.automaticDownloadSupported,
        )
    }

    private fun applyDurationConsensus(candidates: MutableList<RankedSourceCandidate>, song: String) {
        val plausible = candidates.filter {
            !it.previewOnly &&
                it.durationSeconds >= 60.0 &&
                textSimilarity(song, it.title) >= 0.72
        }
        if (plausible.size < 2) return

        val median = median(plausible.map { it.durationSeconds })
        if (median <= 0) return

        candidates.indices.forEach { index ->
            val candidate = candidates[index]
            if (candidate.previewOnly || candidate.durationSeconds <= 0) return@forEach
            val ratio = candidate.durationSeconds / median
            candidates[index] = when {
                ratio < 0.68 -> candidate.copy(
                    score = (candidate.score - 35).coerceAtLeast(0),
                    durationWarning = true,
                    reason = candidate.reason + ", duração muito menor que versões equivalentes (mediana ${durationLabel(median)})",
                )
                ratio > 1.55 -> candidate.copy(
                    score = (candidate.score - 15).coerceAtLeast(0),
                    durationWarning = true,
                    reason = candidate.reason + ", duração muito maior que versões equivalentes (mediana ${durationLabel(median)})",
                )
                ratio in 0.92..1.08 -> candidate.copy(score = (candidate.score + 3).coerceAtMost(100))
                else -> candidate
            }
        }
    }

    private fun median(values: List<Double>): Double {
        if (values.isEmpty()) return 0.0
        val sorted = values.sorted()
        val mid = sorted.size / 2
        return if (sorted.size % 2 == 1) sorted[mid] else (sorted[mid - 1] + sorted[mid]) / 2.0
    }

    private fun levenshtein(a: String, b: String): Int {
        if (a == b) return 0
        if (a.isEmpty()) return b.length
        if (b.isEmpty()) return a.length

        var previous = IntArray(b.length + 1) { it }
        for (i in a.indices) {
            val current = IntArray(b.length + 1)
            current[0] = i + 1
            for (j in b.indices) {
                val cost = if (a[i] == b[j]) 0 else 1
                current[j + 1] = min(
                    min(current[j] + 1, previous[j + 1] + 1),
                    previous[j] + cost,
                )
            }
            previous = current
        }
        return previous[b.length]
    }
}
